sympy/functions/elementary/exponential.py

```diff
@@ -1055,3 +1061,90 @@ def _eval_is_zero(self):
             k = self.args[1]
         if x.is_zero and k.is_zero:
             return True
+
+    def _extract_identity(self):
+        r"""Simplifier using the property
+        `W_{k}\left(z e^{z}\right) = z` for some `k` and `z`.
+
+        Notes
+        =====
+
+        The region where `W_{k}\left(z e^{z}\right)` reduces depends
+        on both the value of `z` and the branch index `k`.
+
+        The boundaries of each regions where
+        `W_{k}\left(z e^{z}\right)` reduces to `z` resembles the
+        quadratrix of Hippias.
+
+        References
+        ==========
+
+        .. [1] https://cs.uwaterloo.ca/research/tr/1993/03/W.pdf
+        """
+        from sympy.functions.elementary.complexes import re, im
+        from sympy.functions.elementary.trigonometric import cot
+
+        args = self.args
+        if len(args) == 1:
+            x, k = args[0], S.Zero
+        else:
+            x, k = args
+
+        z = Wild('z')
+        patt = z * exp(z)
+        match = x.match(patt)
+        if not match:
+            return self
+
+        var = match[z]
+
+        # Skip non-numeric cases
+        if not var.is_number and not k.is_Integer:
+            return self
```

Comment:
Should this be `or` rather than `and`?

Reply:
Maybe I misunderstand but you've changed this from
```
if not var.is_number and not k.is_Integer:
```
to
```
if not (var.is_number or k.is_Integer):
```
which is equivalent.

I was thinking that maybe this should be
```
if not var.is_number or not k.is_Integer:
```
or
```
if not (var.is_number and k.is_Integer):
```
I think we want to exit unless we know that we are in a numeric case which will happen when var and k are *both* numeric.